src/features/devtools/layoutValidation/useLayoutValidation.tsx

```diff
@@ -1,155 +1,21 @@
-            }
-          }
-        } else {
-          out[layoutSetId][pageName] = layout;
-        }
-      }
-    } else {
-      out[layoutSetId] = layouts;
-    }
-  }
-
-  return out;
-}
-
-/**
- * Validates a layout page against the current data model schema (looking up bindings in the schema).
- *
- * You can call this without specifying the repeating groups state, as we'll generate a simple state for you where
- * every repeating group has one row (thus making every possible component appear in the layout).
- */
-function useDataModelBindingsValidation(props: LayoutValidationProps) {
-  const layoutSetId = useCurrentLayoutSetId() || 'default';
-  const { logErrors = false } = props;
-  const schema = useCurrentDataModelSchema();
-  const dataType = useCurrentDataModelType();
-  const nodes = useNodesStructureMemo();
-
-  return useMemo(() => {
-    const failures: LayoutValidationErrors = {
-      [layoutSetId]: {},
-    };
-    if (!schema) {
-      return failures;
-    }
-    const rootElementPath = getRootElementPath(schema, dataType);
-
-    const lookupBinding = (binding: string) =>
-      lookupBindingInSchema({
-        schema,
-        rootElementPath,
-        targetPointer: dotNotationToPointer(binding),
-      });
-
-    for (const [pageName, layout] of Object.entries(nodes.all())) {
-      for (const node of layout.flat(true)) {
-        if ('validateDataModelBindings' in node.def) {
-          const errors = node.def.validateDataModelBindings({
-            node: node as any,
-            lookupBinding,
-          });
-          if (errors.length) {
-            const id = node.item.baseComponentId || node.item.id;
-            failures[layoutSetId][pageName] = failures[layoutSetId][pageName] ?? {};
-            failures[layoutSetId][pageName][id] = errors;
-
-            if (logErrors) {
-              window.logErrorOnce(
-                `Data model binding errors for component '${layoutSetId}/${pageName}/${id}':\n- ${errors.join('\n- ')}`,
-              );
-            }
-          }
-        }
-      }
-    }
-
-    return failures;
-  }, [layoutSetId, schema, dataType, nodes, logErrors]);
-}
-
-/**
- * Utility hook for preventing revalidating bindings unless the node structure actually changes
- * The data model binding validations only depends on the node structure and data model bindings,
- * so it is unecessary to revalidate whenever nodes change for a different reason, and dataModelBindings are static.
- */
-function useNodesStructureMemo() {
-  const nodes = useNodes();
-  const nodesRef = useRef(nodes);
-
-  const allNodesIds = useMemo(() => nodes.allNodes().map((n) => n.item.id), [nodes]);
-  const allNodeIdsRef = useRef(allNodesIds);
-
-  if (allNodesIds === allNodeIdsRef.current || deepEqual(allNodesIds, allNodeIdsRef.current)) {
-    return nodesRef.current;
-  } else {
-    nodesRef.current = nodes;
-    allNodeIdsRef.current = allNodesIds;
-    return nodes;
-  }
-}
```

Comment:
Hvis jeg forstår riktig så er ikke sånne stygge optimaliseringer lenger nødvendig? Fordi man faktisk kan "lytte" til endringer i hierarkiet med en traversal-selector, som ikke påvirkes av alt annet? Sånn som i pagination logikken for å finne ut om en side inneholder valideringsfeil?

I dette tilfelle er vel kanskje hele layout-validerings-logikken flyttet inn i hierarkigenereringen?

Reply:
> Hvis jeg forstår riktig så er ikke sånne stygge optimaliseringer lenger nødvendig?

Korrekt!

> Fordi man faktisk kan "lytte" til endringer i hierarkiet med en traversal-selector, som ikke påvirkes av alt annet?

Ja, korrekt - det vil si, en `traversalSelector()` kjører bare på nytt når det er en endring i antall komponenter (altså noen har blitt lagt til/fjernet). Men en hvilken som helst annen selector vil potensielt kjøre ved hver eneste endring, ja.

> Sånn som i pagination logikken for å finne ut om en side inneholder valideringsfeil?

Du tenker på `useRowStructure()`? Ja, den ble også fjernet her.

> I dette tilfelle er vel kanskje hele layout-validerings-logikken flyttet inn i hierarkigenereringen?

Korrekt. Det meste ligger nå i `src/utils/layout/generator/validation`. Det inkluderer jo også generell feilhåndtering i generatoren, samt feilhåndtering ved rendring av komponenter, så nå vil en exception bare føre til at en gitt komponent tryner.